src/api/uploads/MultiputUpload.js

```diff
@@ -0,0 +1,367 @@
+        this.initialFileLastModified = getFileLastModifiedAsISONoMSIfPossible(this.file);
+
+        this.sessionEndpoints = {
+            createSession: createSessionUrl,
+            uploadPart: null,
+            listParts: null,
+            commit: null,
+            abort: null,
+            logEvent: null
+        };
+        // Set a default logEvent endpoint so that we can report failures from createSession.
+        // This value will be overwritten on successful response from createSession.
+        // This assumes that the createSessionUrl already contains the upload session id.
+        this.sessionEndpoints.logEvent = `${createSessionUrl.replace(/files\/\d+/, 'files')}/log`;
+        // Explicitly cast these two to string because sometimes they get passed as int, and
+        // sending folder_id as a number instead of a string in the JSON body of createSession
+        // results in an API error.
+        this.destinationFolder = destinationFolder;
+        this.destinationFile = destinationFile;
+        this.config = config || DEFAULT_MULTIPUT_CONFIG;
+        this.fileSha1 = null;
+        this.totalUploadedBytes = 0;
+        this.partsNotStarted = 0; // # of parts yet to be processed
+        this.partsDigestComputing = 0; // # of parts sent to the digest worker
+        this.partsDigestReady = 0; // # of parts with digest finished that are waiting to be uploaded.
+        this.partsUploading = 0; // # of parts with upload requests currently inflight
+        this.partsUploaded = 0; // # of parts successfully uploaded
+        this.firstUnuploadedPartIndex = 0; // Index of first part that hasn't been uploaded yet.
+        this.onProgress = null;
+        this.onCompleted = null;
+        this.onFailure = null;
+        this.createSessionNumRetriesPerformed = 0;
+        this.partSize = 0;
+        this.parts = [];
+        this.commitRetryCount = 0;
+        this.ended = false;
+        this.worker = createWorker();
+        this.clientId = null;
+    }
+
+    /**
+     * Creates upload session. If a file ID is supplied, use the Chunked Upload File Version
+     * API to replace the file.
+     *
+     * @private
+     * @param {Object} - Request options
+     * @param {boolean} [options.url] - Upload URL to use
+     * @param {string} [options.fileId] - ID of file to replace
+     * @param {string} [options.fileName] - New name for file - required for new file upload
+     * @return {void}
+     */
+    createUploadSession = async (): Promise<> => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        // Set up post body
+        const postData: StringAnyMap = {
+            file_size: this.file.size,
+            file_name: this.file.name
+        };
+
+        if (this.destinationFolder) {
+            postData.folder_id = this.destinationFolder;
+        }
+
+        try {
+            const data = await this.xhr.post({ url: this.sessionEndpoints.createSessionUrl, data: postData });
```

Comment:
What does this.xhr refer to here?

Reply:
it refers to the Xhr instance created in the Base class it extends from. https://github.com/box/box-ui-elements/blob/master/src/api/Base.js